src/java.base/share/classes/java/lang/StableValue.java

```diff
@@ -0,0 +1,664 @@
+    }
+
+    /**
+     * {@return a new stable function}
+     * <p>
+     * The returned {@link Function function} is a caching function that, for each allowed
+     * input in the given set of {@code inputs}, records the values of the provided
+     * {@code original} function upon being first accessed via the returned function's
+     * {@linkplain Function#apply(Object) apply()} method.
+     * <p>
+     * The provided {@code original} function is guaranteed to be successfully invoked
+     * at most once per allowed input, even in a multi-threaded environment. Competing
+     * threads invoking the returned function's {@linkplain Function#apply(Object) apply()}
+     * method when a value is already under computation will block until a value is
+     * computed or an exception is thrown by the computing thread.
+     * <p>
+     * If the provided {@code original} function throws an exception, it is relayed to
+     * the initial caller and no content is recorded.
+     * <p>
+     * The returned function is not {@link Serializable}.
+     * <p>
+     * If the provided {@code original} function recursively calls the returned
+     * function for the same input, an {@linkplain IllegalStateException} will
+     * be thrown.
+     *
+     * @param inputs   the set of (non-null) allowed input values
+     * @param original Function used to compute cached values
+     * @param <T>      the type of the input to the returned Function
+     * @param <R>      the type of results delivered by the returned Function
+     */
+    static <T, R> Function<T, R> function(Set<? extends T> inputs,
+                                          Function<? super T, ? extends R> original) {
+        Objects.requireNonNull(inputs);
+        Objects.requireNonNull(original);
+        return inputs instanceof EnumSet<?> && !inputs.isEmpty()
+                ? StableEnumFunction.of(inputs, original)
+                : StableFunction.of(inputs, original);
+    }
+
+    /**
+     * {@return a new stable list with the provided {@code size}}
+     * <p>
+     * The returned list is an {@linkplain Collection##unmodifiable unmodifiable} list
+     * whose size is known at construction. The list's elements are computed via the
+     * provided {@code mapper} when they are first accessed
+     * (e.g. via {@linkplain List#get(int) List::get}).
+     * <p>
+     * The provided {@code mapper} int function is guaranteed to be successfully invoked
+     * at most once per list index, even in a multi-threaded environment. Competing
+     * threads accessing an element already under computation will block until an element
+     * is computed or an exception is thrown by the computing thread.
+     * <p>
+     * If the provided {@code mapper} throws an exception, it is relayed to the initial
+     * caller and no value for the element is recorded.
+     * <p>
+     * The returned list and its {@link List#subList(int, int) subList} views implement
```

Comment:
Perhaps the most important fact to note here is that the returned lists via `subList` and friends are _also_ stable lists!